### DISCUSSION TITLE
Building FlatImages from Source with different rootfs & other Questions

### DISCUSSION THREAD

**Discussion by Azathothas** [flatimage/flatimage#34] [General]
> Creating a discussion instead, since I had too many questions and the Issues would have been spammed.

Hi, I have been reading the [deploy.sh](https://github.com/ruanformigoni/flatimage/blob/master/deploy/flatimage.sh) script, and after few hrs finally created my own homegrown script.
I had some questions, especially about the [_create_elf function](https://github.com/ruanformigoni/flatimage/blob/master/deploy/flatimage.sh#L79)

1. My homegrown script kept failing and eventually it turns out I must only include that exact list of binaries:
```bash
for binary in bin/{bash,busybox,bwrap,ciopfs,dwarfs_aio,fim_portal,fim_portal_daemon,fim_bwrap_apparmor,janitor,lsof,overlayfs,proot}; do
```
Or maybe it could have been that I was also including the `boot` binary too.
Is that list really hardcoded...? Or was it the presence of the `boot` binary that caused the error: `Invalid dwarfs system`

2. I wanted to check out if musl was supported, I ran the steam.flatimage (runs perfectly on glibc distros) on alpine, but it didn't launch.
Is that a known bug....?
My logs are here: https://paste.ajam.dev/p/MYau2I

**Comment by ruanformigoni** [comment]
Hello! 

> 1. My homegrown script kept failing and eventually it turns out I must only include that exact list of binaries:

You an increase/decrease the list of binaries, but they need to match the ones in here, in the same order, adjust accordingly:
https://github.com/ruanformigoni/flatimage/blob/6bc6815548d136d10c6fe3b6bd254e9f28beb599/src/boot/boot.cpp#L194-L211

---

> I wanted to check out if musl was supported, I ran the steam.flatimage (runs perfectly on glibc distros) on alpine, but it didn't launch.
Is that a known bug....?

Steam and [gameimage](https://github.com/ruanformigoni/gameimage) worked fine in MUSL from my testing in alpine. The main system being musl should not matter to the glibc container. It seems that something is wrong with the `DISPLAY` variables or permissions to the X server, can you try to use `xhost +` before executing flatimage to see if that fixes it?

**Reply by Azathothas** [reply]
Wow, that worked.
I had to do:
```bash
sudo apk add xhost

$ xhost +
access control disabled, clients can connect from any host

#Relanching steam it works...
```
May I ask why did conty, not fail...?
I tried to use the steam.appimage from: https://github.com/ivan-hc/Steam-appimage/releases/download/continuous/Steam-1.0.0.81-2-3-x86_64.AppImage
And that one launched without any errors.


As for the order of binaries, it would have helped if there was a comment in the deploy script....
Also, would be nice if there was an official packer that accepted a dir to the static binaries, path to the image and did all of that magic automatically...
Thanks for being so helpful.

**Reply by ruanformigoni** [reply]
No problem, I am mainly pursuing to eliminate library and glibc issues with this packaging format. I found a mistake which I made:

https://github.com/ruanformigoni/flatimage/blob/6bc6815548d136d10c6fe3b6bd254e9f28beb599/src/cpp/lib/bwrap.hpp#L156

This linux should be `m_path_dir_xdg_runtime = ns_env::get_or_else("XDG_RUNTIME_DIR", "/run/user/{}"_fmt(getuid()));`. This is probably why this happened, sorry about that. I'll release a patch that also hides the errors from the killed processes.

> Also, would be nice if there was an official packer that accepted a dir to the static binaries, path to the image and did all of that magic automatically...

That would be cool :smile:

**Reply by ruanformigoni** [reply]
Could you please try to do `xhost -` to revert the `xhost +`, confirm that the error still happens. And then do something like `XDG_RUNTIME_DIR=/run/user/$(id -u) ./steam.flatimage`? This would confirm that this caused the issue.

**Reply by Azathothas** [reply]
```bash
$ xhost -
access control enabled, only authorized clients can connect
alpine@alpine-xfce:~$ XDG_RUNTIME_DIR=/run/user/$(id -u) ./steam.flatimage
steam.sh[15741]: Running Steam on arch rolling 64-bit
steam.sh[15741]: STEAM_RUNTIME is enabled automatically
setup.sh[15811]: Steam runtime environment up-to-date!
steam.sh[15741]: Steam client's requirements are satisfied
[2024-10-18 13:40:26] Startup - updater built Sep 17 2024 20:02:19
[2024-10-18 13:40:26] Startup - Steam Client launched with: '/home/steam/.local/share/Steam/ubuntu12_32/steam' '-srt-logger-opened'
10/18 13:40:26 minidumps folder is set to /tmp/dumps
10/18 13:40:26 Init: Installing breakpad exception handler for appid(steam)/version(1726604483)/tid(15858)
Authorization required, but no authorization protocol specified

src/steamexe/updateui_xwin.cpp (341) : Could not open connection to X
src/steamexe/updateui_xwin.cpp (341) : Could not open connection to X
10/18 13:40:26 Init: Installing breakpad exception handler for appid(steam)/version(1726604483)/tid(15858)
Error: Check your DISPLAY environment variable and make sure that you have enabled X.
If you are running remotely, make sure that you have a remote connection which will allow an X connection.

For more information visit https://support.steampowered.com/kb_article.php?ref=4050-WOJB-0608
Using host zenity for message
assert_20241018134026_4.dmp[15863]: Uploading dump (out-of-process)
/tmp/dumps/assert_20241018134026_4.dmp

assert_20241018134026_4.dmp[15863]: Uploading dump (out-of-process)
/tmp/dumps/assert_20241018134026_4.dmp
src/steamexe/main.cpp (1300) : failed to initialize update status ui, or create initial window
src/steamexe/main.cpp (1300) : failed to initialize update status ui, or create initial window
10/18 13:40:26 Init: Installing breakpad exception handler for appid(steam)/version(1726604483)/tid(15858)
assert_20241018134026_7.dmp[15866]: Uploading dump (out-of-process)
/tmp/dumps/assert_20241018134026_7.dmp

assert_20241018134026_7.dmp[15866]: Uploading dump (out-of-process)
/tmp/dumps/assert_20241018134026_7.dmp
Error:
/home/steam/.local/share/Steam/steam_msg.sh: line 74: /usr/bin/cat: Socket not connected
```

**Reply by ruanformigoni** [reply]
Could you check in debug mode if this line is now corrected `I::bwrap.hpp::157::XDG_RUNTIME_DIR: /run/user/{}1000`? I'll install alpine in a VM to investigate further.

**Reply by Azathothas** [reply]
![image](https://github.com/user-attachments/assets/9edeac47-c055-47dd-9fe9-4f319ccbdd0d)

**Reply by ruanformigoni** [reply]
Nice, even if that wasn't the main issue it seems to be fixed now, I'll get back to you on this when I figure it out.

**Reply by ruanformigoni** [reply]
Crazy thing that I setup the machine from scratch and this issue did not happen to me. I only had to install `fuse3` and do `modprobe fuse` for everything to work. Did you use the `setup-desktop` script or install the packages manually?

**Reply by Azathothas** [reply]
Sorry, for causing unnecessary troubles.
I tried to reproduce this on other machines, couldn't. 
So it seems, at some point, I messed up some of my default settings.

As for setup-desktop, I don't think I used that. I probably installed it all manually.
This is an edge case and likely my own system's invalid settings.

**Reply by ruanformigoni** [reply]
Hey no problem, if it happened then there is an edge case somewhere. I'll leave it on my kanban, you also reminded me to update the steam script to handle `busybox wget` which does not support some options, and use `sh` instead of `bash` to not have to install it on alpine.

**Reply by Azathothas** [reply]
I figured out what was happening.
I use rdp/vnc for my sessions and that's why it didn't work.
I tried without rdp, and it launches...

Apparently, conty uses xephyr or something to bypass it completely.

**Reply by ruanformigoni** [reply]
Ah, got it! Maybe it would be interesting to do something like that in here too.

**Reply by Azathothas** [reply]
Hello, again, over the last few days, I setup a repo to build flatimage baseimages with different rootfs/configurations to experiment with over at: https://github.com/Azathothas/flatimage-base

I am learning a lot but I find the documentation a bit...lacking
This is a good intro https://flatimage.github.io/docs/
But there should also be a more detailed version....

I went through your git history and located the old docs:
https://github.com/ruanformigoni/flatimage/blob/62b035968fc47afc0857e955c6b644a8a4c72c5f/README.md

I hope we can expect these to be added to the main docs..when you have time.
Thanks again for this awesome work, [my friends and I](https://t.me/official_loonix/3966) have been testing a lot of portable formats and yours has impressed us the most.

**Reply by ruanformigoni** [reply]
> Hello, again, over the last few days, I setup a repo to build flatimage baseimages with different rootfs/configurations to experiment with over at: https://github.com/Azathothas/flatimage-base

Really cool to see other distributions work with flatimage, I have limited to `arch` and `alpine` to have one gnu and one musl alternative and to ease effort to maintain more than two distributions.

> I am learning a lot but I find the documentation a bit...lacking
This is a good intro https://flatimage.github.io/docs/
But there should also be a more detailed version....

Yes, I need to include more examples in the documentation.

> Thanks again for this awesome work, [my friends and I](https://t.me/official_loonix/3966) have been testing a lot of portable formats and yours has impressed us the most.

No problem, it is really nice to see that this effort was useful for someone else as well. Do you have a discord? I would prefer it since I do not use telegram often.

**Reply by Azathothas** [reply]
Yes, I repurposed your steam script to use cachyos, and I noticed speed gains in terms of startup and also smaller size of the final image.

If you could include an example to bundle something like distrobox on alpine:
- Would show me how to handle apps that need other apps, for example distrobox needs podman (rootless) to work

And another one, signal-desktop, on arch:
- Would show me how to handle electron apps.


We moved to matrix because file limit is there `100 MB` and then we moved to telegram because file limit is `2GB`
And for working on portable apps, this was really important since on average, the size usually exceeds 100MB easily...
